File: test/python/bench/test_qft_adder.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the qft_adder benchmark."""

from __future__ import annotations

import json

import pytest

from mqt.core.bench import qft_adder

from .utils import assert_generates


@pytest.mark.parametrize("method", [qft_adder.Method.REGISTER, qft_adder.Method.CONSTANT])
@pytest.mark.parametrize("overflow", [qft_adder.Overflow.WRAP, qft_adder.Overflow.CARRY])
def test_qft_adder_reference_json_and_generation(method: qft_adder.Method, overflow: qft_adder.Overflow) -> None:
    """Expose both operand representations with the same overflow contract."""
    benchmark = qft_adder.QFTAdder(qft_adder.Options(addend="110", accumulator="011", method=method, overflow=overflow))
    expected_sum = "1001" if overflow == qft_adder.Overflow.CARRY else "001"
    expected = ("110" if method == qft_adder.Method.REGISTER else "") + expected_sum
    assert benchmark.options.addend == "110"
    assert benchmark.options.accumulator == "011"
    assert benchmark.output.width == len(expected)
    assert benchmark.expected_result == expected
    assert benchmark.probability(expected) == 1
    assert benchmark.evaluate({expected: 8}).success_probability == 1
    parameters = json.loads(benchmark.instance_specification_json)["parameters"]
    assert parameters["addend"] == "110"
    assert parameters["accumulator"] == "011"
    copy = qft_adder.QFTAdder.from_instance_specification_json(benchmark.instance_specification_json)
    manifest_copy = qft_adder.QFTAdder.from_manifest_json(benchmark.manifest_json)
    assert copy.case_id == manifest_copy.case_id == benchmark.case_id
    assert_generates(benchmark.generate())


def test_qft_adder_superposition_reference() -> None:
    """Keep the observable correlation for a partly superposed addend."""
    benchmark = qft_adder.QFTAdder(qft_adder.Options(addend="1+0", accumulator="001"))
    assert benchmark.expected_result is None
    assert benchmark.probability("100101") == pytest.approx(0.5)
    assert benchmark.probability("110111") == pytest.approx(0.5)
    assert benchmark.probability("000001") == 0
    assert benchmark.probability("100100") == 0
    evaluation = benchmark.evaluate({"100101": 1, "110111": 1})
    assert evaluation.total_variation_distance == 0
    assert evaluation.success_probability is None

File: test/bench/test_qft_adder.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/QFTAdder.hpp"

#include "support/Diagnostics.hpp"
#include "support/TestSupport.hpp"

#include "gtest/gtest.h"

#include <string>

namespace {

using mqt::bench::QFTAdder;
using mqt::bench::QFTAdderMethod;
using mqt::bench::QFTAdderOptions;
using mqt::bench::QFTAdderOverflow;

TEST(QFTAdder, PreservesConfiguredOperandsAndOverflow) {
  for (const auto method :
       {QFTAdderMethod::Register, QFTAdderMethod::Constant}) {
    for (const auto overflow :
         {QFTAdderOverflow::Wrap, QFTAdderOverflow::Carry}) {
      const auto benchmark = ::mqt::test::value(QFTAdder::create({
          .addend = "011",
          .accumulator = "110",
          .method = method,
          .overflow = overflow,
      }));
      const auto* const sum =
          overflow == QFTAdderOverflow::Carry ? "1001" : "001";
      const auto expected =
          (method == QFTAdderMethod::Register ? std::string{"011"}
                                              : std::string{}) +
          sum;
      EXPECT_EQ(benchmark.options().addend, "011");
      EXPECT_EQ(benchmark.options().accumulator, "110");
      EXPECT_EQ(benchmark.output().width, expected.size());
      EXPECT_EQ(benchmark.expectedResult(), expected);
      EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability(expected)), 1.);
      EXPECT_EQ(::mqt::test::value(benchmark.evaluate({{expected, 16}}))
                    .successProbability,
                1.);
    }
  }
}

TEST(QFTAdder, KeepsLeadingZerosAndRejectsUnsupportedInputs) {
  const auto zero = ::mqt::test::value(QFTAdder::create({
      .addend = "000",
      .accumulator = "000",
      .method = QFTAdderMethod::Constant,
      .overflow = QFTAdderOverflow::Carry,
  }));
  EXPECT_EQ(zero.expectedResult(), "0000");
  for (const auto& options : {
           QFTAdderOptions{.addend = "", .accumulator = ""},
           QFTAdderOptions{.addend = "01", .accumulator = "1"},
           QFTAdderOptions{.addend = "x", .accumulator = "0"},
           QFTAdderOptions{.addend = "1", .accumulator = "+"},
           QFTAdderOptions{
               .addend = "+",
               .accumulator = "0",
               .method = QFTAdderMethod::Constant,
           },
       }) {
    EXPECT_EQ(::mqt::test::errorKind([&] { return QFTAdder::create(options); }),
              ::mqt::ErrorCategory::InvalidArgument);
  }
  EXPECT_EQ(::mqt::test::errorKind([&] { return zero.probability("000"); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] { return zero.probability("000x"); }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(QFTAdder, ScoresTheCorrelatedSuperposition) {
  const auto benchmark = ::mqt::test::value(
      QFTAdder::create({.addend = "1+0", .accumulator = "001"}));
  EXPECT_FALSE(benchmark.expectedResult());
  EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("100101")), 0.5);
  EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("110111")), 0.5);
  EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("000001")), 0.);
  EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("100100")), 0.);
  const auto exact =
      ::mqt::test::value(benchmark.evaluate({{"100101", 8}, {"110111", 8}}));
  EXPECT_DOUBLE_EQ(exact.totalVariationDistance, 0.);
  EXPECT_DOUBLE_EQ(exact.squaredHellingerFidelity, 1.);
  EXPECT_FALSE(exact.successProbability);
  const auto biased = ::mqt::test::value(benchmark.evaluate({{"100101", 16}}));
  EXPECT_DOUBLE_EQ(biased.totalVariationDistance, 0.5);
  EXPECT_DOUBLE_EQ(biased.squaredHellingerFidelity, 0.5);
}

TEST(QFTAdder, BoundsTheSumWidthAndKeepsReferenceWeightsRepresentable) {
  const auto width = QFTAdderOptions::MAX_QUBITS;
  const auto accumulator = std::string(width - 1, '0') + "1";
  const auto maximum = ::mqt::test::value(QFTAdder::create(
      {.addend = std::string(width, '+'), .accumulator = accumulator}));
  EXPECT_GT(::mqt::test::value(maximum.probability(std::string(width, '1') +
                                                   std::string(width, '0'))),
            0.);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return QFTAdder::create({.addend = std::string(width, '1'),
                                       .accumulator = accumulator,
                                       .overflow = QFTAdderOverflow::Carry});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_NO_THROW(static_cast<void>(::mqt::test::value(
      QFTAdder::create({.addend = std::string(width - 1, '1'),
                        .accumulator = accumulator.substr(1),
                        .overflow = QFTAdderOverflow::Carry}))));
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return QFTAdder::create(
                  {.addend = std::string(width + 1, '0'),
                   .accumulator = std::string(width + 1, '0')});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
}

} // namespace
